src/profiler.rs

```diff
@@ -137,10 +141,19 @@ impl ProfilerGuardBuilder {
                 }
 
                 match profiler.start() {
-                    Ok(()) => Ok(ProfilerGuard::<'static> {
-                        profiler: &PROFILER,
-                        timer: Some(Timer::new(self.frequency)),
-                    }),
+                    Ok(()) => {
+                        #[cfg(target_os = "linux")]
+                        {
+                            let running_arc = profiler.running.clone();
+                            std::thread::spawn(move || {
+                                fire_rt_signals(running_arc, self.frequency as u64)
+                            });
```

Comment:
I believe there is a race happening here.
Consider this sequence of events:

- you start the profiler.
- the signalling thread runs and starts its main loop because `running` is `true`.
- the signalling thread does a `sleep`.
- during that time, you stop the profiler.
- you restart the profiling, spawning a *second* signalling thread.
- the first signalling thread wakes up from `sleep` and continues.

Now you have two signalling threads. I believe there is a simple solution to this:
Create a fresh `running` `Arc` when you stop the profiler. That way the currently running signalling thread will not share the same flag with a newly started profiling session.

Reply:
good catch indeed!